src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::usize;
+use std::sync::{Arc, Mutex, Condvar};
+use std::thread::{Builder, JoinHandle};
+use std::boxed::FnBox;
+use std::collections::{BinaryHeap, HashMap, VecDeque};
+use std::cmp::Ordering;
+use std::sync::atomic::{AtomicUsize, Ordering as AtomicOrdering};
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+pub struct Task<T> {
+    // The task's id in the pool. Each task has a unique id,
+    // and it's always bigger than preceding ones.
+    id: u64,
+    // which group the task belongs to.
+    group_id: T,
+    task: Box<FnBox() + Send>,
+}
+
+impl<T> Task<T> {
+    fn new<F>(id: u64, group_id: T, job: F) -> Task<T>
+        where F: FnOnce() + Send + 'static
+    {
+        Task {
+            id: id,
+            group_id: group_id,
+            task: Box::new(job),
+        }
+    }
+}
+
+impl<T> Ord for Task<T> {
+    fn cmp(&self, right: &Task<T>) -> Ordering {
+        self.id.cmp(&right.id).reverse()
+    }
+}
+
+impl<T> PartialEq for Task<T> {
+    fn eq(&self, right: &Task<T>) -> bool {
+        self.cmp(right) == Ordering::Equal
+    }
+}
+
+impl<T> Eq for Task<T> {}
+
+impl<T> PartialOrd for Task<T> {
+    fn partial_cmp(&self, rhs: &Task<T>) -> Option<Ordering> {
+        Some(self.cmp(rhs))
+    }
+}
+
+pub trait ScheduleQueue<T> {
+    fn pop(&mut self) -> Option<Task<T>>;
+    fn push(&mut self, task: Task<T>);
+    fn on_task_finished(&mut self, group_id: &T);
+    fn on_task_started(&mut self, group_id: &T);
+}
+
+pub struct PushError<T>(pub T);
+
+// `BigGroupThrottledQueue` tries to throttle group's concurrency to
+//  `group_concurrency_on_busy` when it's busy.
+// When one worker asks a task to run, it schedules in the following way:
+// 1. Find out which group has a running number that is smaller than
+//    that of `group_concurrency_on_busy`.
+// 2. If more than one group meets the first point, run the one who
+//    comes first.
+// If no group meets the first point, choose according to the following rules:
+// 1. Select the groups with least running tasks.
+// 2. If more than one group meets the first point,choose the one
+//     whose task comes first(with the minimum task's ID)
+pub struct BigGroupThrottledQueue<T> {
```

Comment:
I still prefer to use `BigGroupThrottledQueue ` instead of `FairQueue ` since the queue will throttle big group's concurrency when server is busy. And `BigGroupThrottledQueue ` is more clear while `FairQueue` is a bit fuzzy @hhkbp2